processor/lsmintervalprocessor/internal/merger/value.go

```diff
@@ -0,0 +1,796 @@
+		merge(
+			m.Summary().DataPoints(),
+			metricID,
+			v.addSummaryDataPoint,
+			// Assume summary to be cumulative temporality
+			pmetric.AggregationTemporalityCumulative,
+		)
+	case pmetric.MetricTypeHistogram:
+		merge(
+			m.Histogram().DataPoints(),
+			metricID,
+			v.addHistogramDataPoint,
+			m.Histogram().AggregationTemporality(),
+		)
+	case pmetric.MetricTypeExponentialHistogram:
+		merge(
+			m.ExponentialHistogram().DataPoints(),
+			metricID,
+			v.addExponentialHistogramDataPoint,
+			m.ExponentialHistogram().AggregationTemporality(),
+		)
+	}
+}
+
+func (s *Value) getOverflowResourceBucketID() (identity.Resource, error) {
+	r := pcommon.NewResource()
+	if err := decorate(
+		r.Attributes(),
+		s.resourceLimitCfg.Overflow.Attributes,
+	); err != nil {
+		return identity.Resource{}, fmt.Errorf("failed to create overflow bucket: %w", err)
+	}
+	return identity.OfResource(r), nil
+}
+
+func (s *Value) getOverflowScopeBucketID(
+	res identity.Resource,
+) (identity.Scope, error) {
+	scope := pcommon.NewInstrumentationScope()
+	if err := decorate(
+		scope.Attributes(),
+		s.scopeLimitCfg.Overflow.Attributes,
+	); err != nil {
+		return identity.Scope{}, fmt.Errorf("failed to create overflow bucket: %w", err)
+	}
+	return identity.OfScope(res, scope), nil
+}
+
+type dataPointSlice[DP dataPoint[DP]] interface {
+	Len() int
+	At(i int) DP
+	AppendEmpty() DP
+}
+
+type dataPoint[Self any] interface {
+	pmetric.NumberDataPoint | pmetric.SummaryDataPoint | pmetric.HistogramDataPoint | pmetric.ExponentialHistogramDataPoint
+
+	Timestamp() pcommon.Timestamp
+	SetTimestamp(pcommon.Timestamp)
+	Attributes() pcommon.Map
+	CopyTo(dest Self)
+}
+
+func merge[DPS dataPointSlice[DP], DP dataPoint[DP]](
+	from DPS,
+	toMetricID identity.Metric,
+	addDP func(identity.Metric, DP) (DP, bool),
+	temporality pmetric.AggregationTemporality,
+) {
+	switch temporality {
+	case pmetric.AggregationTemporalityCumulative:
+		mergeCumulative(from, toMetricID, addDP)
+	case pmetric.AggregationTemporalityDelta:
+		mergeDelta(from, toMetricID, addDP)
+	}
+}
+
+func mergeCumulative[DPS dataPointSlice[DP], DP dataPoint[DP]](
+	from DPS,
+	toMetricID identity.Metric,
+	addDP func(identity.Metric, DP) (DP, bool),
+) {
+	var zero DP
+	for i := 0; i < from.Len(); i++ {
+		fromDP := from.At(i)
+		toDP, ok := addDP(toMetricID, fromDP)
+		if toDP == zero {
+			// Overflow, discard the datapoint
+			continue
+		}
+		if ok || fromDP.Timestamp() > toDP.Timestamp() {
+			fromDP.CopyTo(toDP)
+		}
+	}
+}
+
+func mergeDelta[DPS dataPointSlice[DP], DP dataPoint[DP]](
+	from DPS,
+	toMetricID identity.Metric,
+	addDP func(identity.Metric, DP) (DP, bool),
+) {
+	var zero DP
+	for i := 0; i < from.Len(); i++ {
+		fromDP := from.At(i)
+		toDP, ok := addDP(toMetricID, fromDP)
+		if toDP == zero {
+			// Overflow, discard the datapoint
+			continue
+		}
+		if ok {
+			// New data point is created so we can copy the old data directly
+			fromDP.CopyTo(toDP)
+			continue
+		}
```

Comment:
Comparing to zero feels like an unnecessarily expensive way to do this. See my other comment about returning true only when we should update. Then this could become

```go
if toDP, exists := addDP(toMetricID, fromDP); exists {
	switch fromDP := any(fromDP).(type) {
	case pmetric.NumberDataPoint:
		mergeDeltaSumDP(fromDP, any(toDP).(pmetric.NumberDataPoint))
	case pmetric.HistogramDataPoint:
		mergeDeltaHistogramDP(fromDP, any(toDP).(pmetric.HistogramDataPoint))
	case pmetric.ExponentialHistogramDataPoint:
		mergeDeltaExponentialHistogramDP(fromDP, any(toDP).(pmetric.ExponentialHistogramDataPoint))
	}
}
```